java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java

```diff
@@ -0,0 +1,245 @@
+      scan(tree.thenStatement());
+      nesting--;
+      boolean elseStatementNotIF = tree.elseStatement() != null && !tree.elseStatement().is(IF_STATEMENT);
+      if(elseStatementNotIF) {
+        increaseComplexity(tree.elseKeyword(), nesting);
+        nesting++;
+      }
+      scan(tree.elseStatement());
+      if(elseStatementNotIF) {
+        nesting--;
+      }
+    }
+
+    @Override
+    public void visitTryStatement(TryStatementTree tree) {
+      scan(tree.resources());
+      scan(tree.block());
+      tree.catches().forEach(c -> increaseComplexity(c.catchKeyword(), nesting));
+      nesting++;
+      scan(tree.catches());
+      nesting--;
+      scan(tree.finallyBlock());
+    }
+
+    @Override
+    public void visitForStatement(ForStatementTree tree) {
+      increaseComplexity(tree.forKeyword(), nesting);
+      nesting++;
+      super.visitForStatement(tree);
+      nesting--;
+    }
+
+    @Override
+    public void visitForEachStatement(ForEachStatement tree) {
+      increaseComplexity(tree.forKeyword(), nesting);
+      nesting++;
+      super.visitForEachStatement(tree);
+      nesting--;
+    }
+
+    @Override
+    public void visitWhileStatement(WhileStatementTree tree) {
+      increaseComplexity(tree.whileKeyword(), nesting);
+      nesting++;
+      super.visitWhileStatement(tree);
+      nesting--;
+    }
+
+    @Override
+    public void visitDoWhileStatement(DoWhileStatementTree tree) {
+      increaseComplexity(tree.doKeyword(), nesting);
+      nesting++;
+      super.visitDoWhileStatement(tree);
+      nesting--;
+    }
+
+    @Override
+    public void visitConditionalExpression(ConditionalExpressionTree tree) {
+      increaseComplexity(tree.questionToken(), nesting);
+      nesting++;
+      super.visitConditionalExpression(tree);
+      nesting--;
+    }
+
+    @Override
+    public void visitSwitchStatement(SwitchStatementTree tree) {
+      increaseComplexity(tree.switchKeyword(), nesting);
+      nesting++;
+      super.visitSwitchStatement(tree);
+      nesting--;
+    }
+
+    @Override
+    public void visitBreakStatement(BreakStatementTree tree) {
+      if(tree.label() != null) {
+        increaseComplexity(tree.breakKeyword(), 1);
+      }
+      super.visitBreakStatement(tree);
+    }
+
+    @Override
+    public void visitContinueStatement(ContinueStatementTree tree) {
+      if(tree.label() != null) {
+        increaseComplexity(tree.continueKeyword(), 1);
+      }
+      super.visitContinueStatement(tree);
+    }
+
+    @Override
+    public void visitClass(ClassTree tree) {
+      nesting++;
+      super.visitClass(tree);
+      nesting--;
+    }
+
+    @Override
+    public void visitLambdaExpression(LambdaExpressionTree lambdaExpressionTree) {
+      nesting++;
+      super.visitLambdaExpression(lambdaExpressionTree);
+      nesting--;
+    }
+
+    @Override
+    public void visitBinaryExpression(BinaryExpressionTree tree) {
+      if(tree.is(CONDITIONAL_AND, CONDITIONAL_OR)) {
```

Comment:
I would also consider the `XOR` (`^`) operator when the expression is a boolean expression.